### PR TITLE
Clear update registry when mutable state failover version changes

### PR DESCRIPTION
**Pull request by alexshtin**
## What changed?

Clear update registry when mutable state failover version changes.

## Why?

Update registry must stay in sync with mutable state. If mutable state was changed because of fail over, update registry must be updated. The easiest way to do it, is clear registry and rebuild from new version of mutable state.

## How did you test it?

Existing tests.

## Potential risks

No risks.

## Documentation

No.

## Is hotfix candidate?

No.

### CODE REVIEW

File: service/history/workflow/context.go [hunk 1]
```diff
@@ -915,13 +915,20 @@ func (c *ContextImpl) ReapplyEvents(
...
 // TODO: remove `ms` parameter again (added since it's not possible to initialize a new Context with a specific MutableState)
 func (c *ContextImpl) UpdateRegistry(ctx context.Context, ms MutableState) update.Registry {
+	var ns *namespace.Namespace
+	if c.MutableState != nil {
+		ns = c.MutableState.GetNamespaceEntry()
+	} else {
+		ns = ms.GetNamespaceEntry()
+	}
+
+	if c.updateRegistry != nil && c.updateRegistry.FailoverVersion() != ns.FailoverVersion() {
```
Comment (yycptt) [inline service/history/workflow/context.go:925]: At least now, the GetCurrentVersion from mutable state will stop updating after workflow is closed, while ns.FailoverVersion will keep changing.  I think that's probably not something you want?

Comment (alexshtin) [reply service/history/workflow/context.go:925]: I definitely don't want to compare fields which are always different after WF is closed :-) I will use namespace version for both.

File: service/history/workflow/context.go [hunk 2]
```diff
@@ -919,13 +919,26 @@ func (c *ContextImpl) ReapplyEvents(
 
 // TODO: remove `ms` parameter again (added since it's not possible to initialize a new Context with a specific MutableState)
 func (c *ContextImpl) UpdateRegistry(ctx context.Context, ms MutableState) update.Registry {
+	var currentVersion int64
+	if c.MutableState != nil {
+		currentVersion = c.MutableState.GetCurrentVersion()
+	} else {
+		currentVersion = ms.GetCurrentVersion()
+	}
```
Comment (dandavison) [inline service/history/workflow/context.go:927]: I'll trust you on this logic! Hopefully one day we can reduce the number of references to MS...

File: service/history/workflow/context.go [hunk 3]
```diff
@@ -919,13 +919,26 @@ func (c *ContextImpl) ReapplyEvents(
...
+
+	if c.updateRegistry != nil && c.updateRegistry.FailoverVersion() != currentVersion {
```
Comment (dandavison) [inline service/history/workflow/context.go:929]: Beautiful.

File: service/history/workflow/context.go [hunk 4]
```diff
@@ -938,17 +951,15 @@ func (c *ContextImpl) UpdateRegistry(ctx context.Context, ms MutableState) updat
...
 				func() int {
-					return c.config.WorkflowExecutionMaxTotalUpdates(nsIDStr)
+					return c.config.WorkflowExecutionMaxTotalUpdates(nsID.String())
 				},
 			),
 		)
-	} else {
-		c.updateRegistry.UpdateFromStore()
 	}
 	return c.updateRegistry
```
Comment (dandavison) [inline service/history/workflow/context.go:964]: Let's switch to this cleaned up version:

```go
func (c *ContextImpl) UpdateRegistry(ctx context.Context, fallbackMutableState MutableState) update.Registry {
	ms := c.MutableState
	if ms == nil {
		ms = fallbackMutableState
	}
	currentVersion := ms.GetCurrentVersion()

	if c.updateRegistry != nil && c.updateRegistry.FailoverVersion() != currentVersion {
		c.updateRegistry.Clear()
		c.updateRegistry = nil
	}

	if c.updateRegistry == nil {
		nsID := ms.GetNamespaceEntry().ID()

		c.updateRegistry = update.NewRegistry(
			func() update.Store { return ms },
			update.WithLogger(c.logger),
			update.WithMetrics(c.metricsHandler),
			update.WithTracerProvider(trace.SpanFromContext(ctx).TracerProvider()),
			update.WithInFlightLimit(
				func() int {
					return c.config.WorkflowExecutionMaxInFlightUpdates(nsID.String())
				},
			),
			update.WithTotalLimit(
				func() int {
					return c.config.WorkflowExecutionMaxTotalUpdates(nsID.String())
				},
			),
		)
	}
	return c.updateRegistry
}
```

**Review by dandavison — APPROVED** [review]
Excellent! Nice work.

I'd like to simplify the `UpdateRegistry` function as indicated, by doing the MS reference choice once. But other from that, LGTM; I believe we have sufficient test coverage for this change in the xdc replication tests.

**Review by yycptt — APPROVED** [review]

**Pull request merged**